From 2a77855d019072f21c9b288a71213454316ef696 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Fri, 28 Oct 2022 16:30:34 +0300
Subject: [PATCH] [SW-109230] gaudi3: replace memset with for loop

Currently memset is used to initialize 3KB of DCCM as part
of ROM init. However memset uses prefetch instruction which
perform reads from DCCM before writing it. To avoid read before
write this patch make use of local copy mechanism to initialize
first 3KB of memory.

Change-Id: I37e049c776f2a9558459989515bd786c16c1d895
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index ed37837..bd0bd84 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -463,13 +463,17 @@
 {
 	uint8_t die_id = base_readl(HL_CFG_BASE + mmD0_PARC_LOCAL_CFG_DIE_ID);
 	uint32_t die_cold_rst_flop_2 = HL_CFG_BASE + mmCOLD_RST_FLOPS_2 + die_id * HL_CFG_SPACE_DIE1_OFFSET;
+	volatile uint32_t *dccm_addr = (uint32_t *) HL_DCCM_BASE;
+	uint32_t dccm_dwords = HL_ARC1_DCCM_PERSIST_SECTION_SIZE >> 2;
 	struct gaudi3_cold_rst_data cold_rst_data;
+	uint32_t i;
 
 	cold_rst_data.data = base_readl(die_cold_rst_flop_2);
 	if (cold_rst_data.arc1_persist_dccm_init_done)
 		goto non_zero_init;
 
-	memset((void *)HL_DCCM_BASE, 0, HL_ARC1_DCCM_PERSIST_SECTION_SIZE);
+	for (i = 0; i < dccm_dwords; i++)
+		*dccm_addr++ = 0;
 
 	cold_rst_data.arc1_persist_dccm_init_done = 1;
 	base_writel(cold_rst_data.data, die_cold_rst_flop_2);
